### PR TITLE
Add support for "spring.kafka.listener.batch-listener" property 

### PR DESCRIPTION
**Pull request by mzagar**
With this feature `ConcurrentKafkaListenerContainerFactoryConfigurer` is able to set batch listener option on `listenerContainerFactory` out of the box if so specified in kafka configuration properties option.

This enables me to write something like this which will work out of the box without any additional configuration (currently this is not the case and I need to create my own instance of `ConcurrentKafkaListenerContainerFactory`; overriding `CustomConcurrentKafkaListenerContainerFactoryConfigurer` is also not an option since it seems this class was not ment to be extended):
```
@SpringBootApplication
@EnableKafka
open class Application {
    val LOG = LoggerFactory.getLogger(this.javaClass.name)

    companion object {
        init {
            System.setProperty("spring.kafka.bootstrapServers", "kafka:9092");
            System.setProperty("spring.kafka.consumer.enableAutoCommit", "false");
            System.setProperty("spring.kafka.consumer.fetchMinSize", (10 * 1024 * 1024).toString());
            System.setProperty("spring.kafka.consumer.fetchMaxWait", 5000.toString());
            System.setProperty("spring.kafka.consumer.clientId", "mzagar-spring-kafka-test");
            System.setProperty("spring.kafka.consumer.groupId", "mzagar-spring-kafka-test");
            System.setProperty("spring.kafka.consumer.maxPollRecords", 5000.toString());
            System.setProperty("spring.kafka.listener.ackMode", MANUAL.name);
            System.setProperty("spring.kafka.listener.concurrency", "10");
            
            // without this I need to manually define ConcurrentKafkaListenerContainerFactory to set batchListener flag
            System.setProperty("spring.kafka.listener.batchListener", "true");
        }
    }

    @KafkaListener(id = "me", topics = arrayOf("my-topic"))
    fun listen(cr: List<ConsumerRecord<String,String>>, ack: Acknowledgment) {
        LOG.info(Thread.currentThread().name + " --> size=" + cr.size)
        ack.acknowledge()
    }
}

fun main(args: Array<String>) {
    SpringApplication.run(Application::class.java, *args)
}
```

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@mzagar Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=9448)!

**Comment by mzagar** [comment]
@garyrussell - I would appreciate if you could take a look and give me feedback if there already exist an easier or 'proper' way to enable batch listener since I expect this to work out of the box but did not. thx!

**Comment by garyrussell** [comment]
@mzagar  This looks like a reasonable update to me; however, it also needs an [asciidoc properties update](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc).

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaProperties.java [hunk 1]
```diff
@@ -672,6 +672,11 @@ public void setDefaultTopic(String defaultTopic) {
 		 */
 		private Long ackTime;
 
+		/**
+		 * True if this should be batch listener.
+		 */
+		private Boolean batchListener;
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaProperties.java]: What is the advantage of setting this to `null` by default? If there is a reasonable default that isn't likely to change, I'd rather have a `boolean` with the default value set. That way you can discover in your IDE if that's enabled by default or not.

File: spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc [hunk 2]
```diff
@@ -986,6 +986,7 @@ content into your application; rather pick only the properties that you need.
 	spring.kafka.listener.ack-time= # Time in milliseconds between offset commits when ackMode is "TIME" or "COUNT_TIME".
 	spring.kafka.listener.concurrency= # Number of threads to run in the listener containers.
 	spring.kafka.listener.poll-timeout= # Timeout in milliseconds to use when polling the consumer.
+	spring.kafka.listener.batch-listener= # If true listener container factory will be configured to create batch listeners.
```
Comment (snicoll) [inline spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc]: Documentation here matches exactly the field javadoc. If you had to change it here, it is a good indicator that the one in the metadata can be improved as well.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaProperties.java (hunk 1 above)
Comment (mzagar) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaProperties.java]: I did not wan't to potentially change any existing behaviour and also saw that all other properties in ConcurrentKafkaListenerContainerFactoryConfigurer are handled as nullable so I just used the same approach here.

Thinking about it now, I don't see a reason why it shouldn't be set to 'false' - this is effectively the default behavior currently.

Comment (mzagar) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/kafka/KafkaProperties.java]: Added this change as [separate commit](https://github.com/spring-projects/spring-boot/pull/9448/commits/daaa4da46b363477aba705a31af2c17740c2a714), let me know what you think. Thanks!

**Comment by snicoll** [comment]
@garyrussell I was reviewing that PR once more and that `batch-listener` boolean property is really about to figure out if a `Record` or `Batch` listener has to be created. A boolean flag looks a bit odd to me. Isn't a listener type or something a better call for this? What if there is a third option at some point? 

This is more an issue in Spring Kafka but I wanted your thoughts before moving forward: exposing a boolean property that changes the listener type is something unusual.

**Comment by garyrussell** [comment]
@snicoll I think a third (or more) flavor is rather unlikely.

Under the covers, when we `poll()` the `KafkaConsumer` it returns a `ConsumerRecords` object which is a container for one or more `ConsumerRecord`.

The default behavior of the container is to spoon feed each record to the listener one-at-a-time. When `batchListener` is `true`, we present a `List<ConsumerRecord>` to the listener - i.e. all the contents of the `ConsumerRecords` as a "batch".

**Comment by wilkinsona** [comment]
FWIW, it sounds to me like an enum would be more descriptive, even if we just map it onto the existing boolean

**Comment by snicoll** [comment]
Agreed, I went ahead and created a `spring.kafka.listener.type` property that can be `single` or `batch`.

**Pull request closed without merging**